### ISSUE TITLE
onDidReceiveDebugSessionCustomEvent comes before onDidStartDebugSession

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#71651]
I rely on `onDidStartDebugSession` to track my debug sessions launching, and then use that information when handling custom events. In current insiders, it seems that the custom events can arrive before `onDidStartDebugSession` has even fired, which means I'm unable to handle them correctly.

I don't know for certain, but I would guess that the start event should always fire first?

```ts
context.subscriptions.push(vs.debug.onDidReceiveDebugSessionCustomEvent((e) => {
			logWarn('onDidReceiveDebugSessionCustomEvent!!!!!!');
});

context.subscriptions.push(vs.debug.onDidStartDebugSession(async (s) => {
			logWarn('onDidStartDebugSession!!!!!!');
});
```

![Screenshot 2019-04-03 at 10 53 40 am](https://user-images.githubusercontent.com/1078012/55470124-ebf0ed00-55fe-11e9-95dc-74e10a46024f.png)

@weinand @isidorn I don't know if this is at all related to the timing issue we saw in a previous version?

**Comment by DanTup** [comment]
The custom events are being sent during handling of `launchRequest` but *after* its response is sent.

```ts
this.sendResponse(response);
this.sendEvent(new Event("dart.launching", { message: "Launching...", finished: false }));
```

**Comment by isidorn** [comment]
This is exactly the reason why we have introduced `onWillStartDebugSession` in the `debugService`. 
@weinand Might it be that we need to call `acceptDebugSessionStarted` in the `mainThreadDebugService` also when a new `onWillStartDebugSession` is fired [here](https://github.com/Microsoft/vscode/blob/0777889acc3b5bdf17e85d7a9c41593ff4d9d38a/src/vs/workbench/api/browser/mainThreadDebugService.ts#L43)

**Comment by DanTup** [comment]
> This is exactly the reason why we have introduced `onWillStartDebugSession` in the debugService.

Is that something for me to use, or something internal? (I can't find any info on it searching on GH here or in the API docs).

**Comment by weinand** [comment]
@DanTup `onWillStartDebugSession` is internal.
I will investigate, but I'm pretty sure that nothing has changed in the last releases.
Are you seeing a regression in the current Insiders?

**Comment by DanTup** [comment]
No, this seems to happen in Stable too. I haven't gone back further to before the previous issue though, though I'm fairly sure it did work at some point in the past (I implementing a progress indicator that this triggers... though now there are additional events that cause it to appear, so it's less obvious it's not appearing on the first event).

**Comment by DanTup** [comment]
@weinand Actually, I think I must've tested badly - this is working fine in Stable, and only the wrong order in Insiders.

Let me know if you want me to make a small repro sample.

**Comment by weinand** [comment]
@DanTup a small repro sample would be great.

**Comment by DanTup** [comment]
Ok, here's a repro:

https://github.com/DanTup/vscode-repro-71651

It's just MockDebug with a minor change to log:

https://github.com/DanTup/vscode-repro-71651/commit/769928975ded0e74b3817bc96f8425226f0e7d22

And it shows this:

<img width="549" alt="Screenshot 2019-04-04 at 4 36 57 pm" src="https://user-images.githubusercontent.com/1078012/55568703-183c6480-56f8-11e9-8a19-a2f4577fc76f.png">

However, it repros in Stable again with this code! So, maybe it's not a regression in the current Insiders, though I don't know if the timing has changed somehow, since in the code I'm testing now for Dart, I can repro really reliably in Insiders, but not in Stable 🤔

**Comment by isidorn** [comment]
You are sending custom events while the session is still not fully initialized.
Those are not guarateed to come after the session started event. The session started event is fired only once the initilized event is done. I believe we had a similar issue in the past, I just can not find it, but I know @DanTup was also involved

**Comment by DanTup** [comment]
@isidorn I'm confused by the above - this issue was opened by me (and might be the one you were looking for). Did you intend to post this on another issue?

FWIW I still think the behaviour is a bit weird - the event is `onDidStart` and not `onDidInitialize`.. if you want to handle custom events that might be sent during initialization, you don't have a reference to the session.

It should at least be called out in the docs though, the description of the event is just "An event which fires when a new debug session has been started". I think this should be more explicit that it's actually fired only when it's fully initialized.

In Dart-Code we queue events if we don't find a matching session, then process them after the session "starts". It feels weird, though it works.

**Comment by isidorn** [comment]
It might be this same issue. Not sure.
Yes I do agree the behaviour is weird and that we should at least improve the docs. We'll see.
